File: 백준/Silver/10816. 숫자 카드 2/숫자 카드 2.py
import sys

N = int(sys.stdin.readline())
arr1 = list(map(int, sys.stdin.readline().split()))
M = int(sys.stdin.readline())
arr2 = list(map(int, sys.stdin.readline().split()))
dict = {}

arr1.sort()

for i in range(len(arr1)):
    if arr1[i] in dict:
        temp = dict.get(arr1[i]) + 1
        dict[arr1[i]] = temp
    else:
        dict[arr1[i]] = 1

for i in range(len(arr2)):
    if arr2[i] in dict:
        print(dict.get(arr2[i]), end=" ")
    else:
        print(0, end=" ")

File: 백준/Silver/10816. 숫자 카드 2/숫자 카드 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {


    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(bf.readLine());
        int[] sang = new int[N];
        String[] sa = bf.readLine().split(" ");
        for (int i = 0; i < N; i++) {
            sang[i] = Integer.parseInt(sa[i]);
        }
        Arrays.sort(sang);

        int M = Integer.parseInt(bf.readLine());
        sa = bf.readLine().split(" ");

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < M; i++) {
            int target = Integer.parseInt(sa[i]);
            int count = upperBound(sang, target) - lowerBound(sang, target);
            sb.append(count).append(" ");
        }

        System.out.println(sb);

    }

    static int lowerBound(int[] arr, int target) {
        int s = 0;
        int e = arr.length - 1;
        while (s <= e) {
            int m = (s + e) / 2;
            if (arr[m] >= target) {
                e = m - 1;
            } else {
                s = m + 1;
            }
        }
        return s;
    }

    static int upperBound(int[] arr, int target) {
        int s = 0;
        int e = arr.length - 1;
        while (s <= e) {
            int m = (s + e) / 2;
            if (arr[m] > target) {
                e = m - 1;
            } else {
                s = m + 1;
            }
        }
        return s;
    }
}